### PR TITLE
Rwc2018/challenge dishwasher2

### PR DESCRIPTION
**Pull request by MatthijsBurgh**

### CODE REVIEW

File: challenge_dishwasher/src/challenge_dishwasher/dishwasher.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+from challenge_dishwasher.simple_grab import FindAndGrab
+from robot_skills.classification_result import ClassificationResult
+from robot_skills.util.entity import Entity
+from robot_smach_states import StartChallengeRobust, NavigateToWaypoint, Say, NavigateToSymbolic
+from robot_smach_states.util import startup
+from robot_smach_states.util.designators import EdEntityDesignator, VariableDesignator, EntityByIdDesignator, Designator
+from robot_smach_states.world_model import Inspect
+from smach import StateMachine, State
+
+STARTING_POINT = "initial_pose"
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/dishwasher.py]: This should be in knowledge

File: challenge_dishwasher/src/challenge_dishwasher/custom_place.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            robot.torso.wait_for_motion_done()
+
+            arm.resolve()._send_joint_trajectory([[-1.1, 0.044, 0.80, 0.297, 0.934, -0.95, 0.4]],
+                                                 timeout=rospy.Duration(0))
+            arm.resolve().wait_for_motion_done()
+
+            goal_pose = PoseStamped()
+            goal_pose.header.stamp = rospy.Time.now()
+            goal_pose.header.frame_id = 'map'
+            goal_pose.pose.position.x = -4.47331285184
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/custom_place.py]: This should be in knowledge

File: challenge_dishwasher/src/challenge_dishwasher/custom_place.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            rospy.sleep(3.0)
+            robot.speech.speak("This is where I usually place my cup")
+            return 'done'
+
+        @cb_interface(outcomes=['done'])
+        def move_away(ud):
+            robot.torso.high()
+            robot.torso.wait_for_motion_done()
+
+            arm.resolve()._send_joint_trajectory([[-1.1, 0.044, 0.80, 0.297, 0.934, -0.95, 0.4]],
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/custom_place.py]: Joint goal should be in knowledge

File: challenge_dishwasher/src/challenge_dishwasher/custom_place.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        arm = Designator(robot.leftArm)
+
+        @cb_interface(outcomes=['done'])
+        def pre_place_pose(ud):
+            robot.torso.high()
+
+            goal_pose = PoseStamped()
+            goal_pose.header.stamp = rospy.Time.now()
+            goal_pose.header.frame_id = 'map'
+            goal_pose.pose.position.x = -4.55638664735
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/custom_place.py]: This should be in knowledge

File: challenge_dishwasher/src/challenge_dishwasher/custom_place.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            goal_pose.pose.position.y = 4.99959353359
+            goal_pose.pose.orientation = Quaternion(*quaternion_from_euler(0, 0, 0.395625992))
+            ControlToPose(robot, goal_pose, ControlParameters(0.5, 1.0, 0.3, 0.3, 0.3, 0.01, 0.1)).execute({})
+
+            robot.torso.wait_for_motion_done()
+            robot.speech.speak('I am now raising my arm')
+
+            arm.resolve()._send_joint_trajectory([[-1.1, 0.044, 0.80, 0.297, 0.934, -0.95, 0.4]],
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/custom_place.py]: Joint goal should be in knowledge

File: challenge_dishwasher/src/challenge_dishwasher/custom_place.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                                                 timeout=rospy.Duration(0))
+            arm.resolve().wait_for_motion_done()
+
+            goal_pose.pose.orientation = Quaternion(*quaternion_from_euler(0, 0, 1.9656259917))
+            ControlToPose(robot, goal_pose, ControlParameters(0.5, 1.0, 0.3, 0.3, 0.3, 0.01, 0.1)).execute({})
+
+            arm.resolve().wait_for_motion_done()
+
+            arm.resolve()._send_joint_trajectory([[-0.92, 0.044, 0.80, 0.497, 0.934, -0.95, 0.4]],
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/custom_place.py]: Joint goal should be in knowledge

File: challenge_dishwasher/src/challenge_dishwasher/open_dishwasher.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+    def _goal_reached(self, dx, dy, dyaw):
+        return math.hypot(dx, dy) < self.params.goal_position_tolerance and abs(dyaw) < self.params.goal_rotation_tolerance
+
+
+class OpenDishwasher(StateMachine):
+    def __init__(self, robot, dishwasher_id):
+        StateMachine.__init__(self, outcomes=['succeeded', 'failed'])
+
+        base_y_position_door = 0
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/open_dishwasher.py]: This should be in knowledge

File: challenge_dishwasher/src/challenge_dishwasher/open_dishwasher.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        base_y_position_rack = -0.75
+        base_y_position_rack2 = -0.65
+        base_yaw_position_rack = 3.7
+
+        @cb_interface(outcomes=['done'])
+        def _pre_grab_handle(ud):
+            robot.rightArm.send_gripper_goal("open", timeout=0)
+            robot.rightArm._send_joint_trajectory([[0, 0.2519052373022729913, 0.7746500794619434, 1.3944848321343395,
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/open_dishwasher.py]: This should be in knowledge

File: challenge_dishwasher/src/challenge_dishwasher/custom_find_cup.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+                if distance(e_p, p) < radius:
+                    return True
+            return False
+
+        msg = rospy.wait_for_message("/amigo/torso_laser/scan", LaserScan)
+        if msg is None:
+            return None
+        msg.header.frame_id = msg.header.frame_id.lstrip('/')
+
+        m_points = Marker()
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/custom_find_cup.py]: I prefer to see the visualization to be seperated from implementation.

File: challenge_dishwasher/src/challenge_dishwasher/custom_find_cup.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+        def distance(p1, p2):
+            return math.hypot(p1[0] - p2[0], p1[1] - p2[1])
+
+        def to_close_to_one_of_these_points(p, points, radius):
+            for e_p in points:
+                if distance(e_p, p) < radius:
+                    return True
+            return False
+
+        msg = rospy.wait_for_message("/amigo/torso_laser/scan", LaserScan)
```
Comment (MatthijsBurgh) [inline challenge_dishwasher/src/challenge_dishwasher/custom_find_cup.py]: I know that this code is developed for AMIGO, but please make it as generic as possible for now.

**Comment by MatthijsBurgh** [comment]
Please make this code more generic. By moving joint goals, position etc to knowledge file. Also remove as much references to AMIGO as possible.

**Comment by MatthijsBurgh** [comment]
@jlunenburg you can approve and merge.

File: challenge_dishwasher/src/challenge_dishwasher/custom_find_cup.py (hunk 10 above)
Comment (reinzor) [reply challenge_dishwasher/src/challenge_dishwasher/custom_find_cup.py]: Better rewrite the whole thing then 😜

**Pull request merged**